### PR TITLE
Multiple movies

### PR DESCRIPTION
**Pull request by alexscheitlin**
### Description

Adds a graphql sub-query to get multiple random movies when fetching a specific movie.

### Related Issues

#28

### Related PR

This work is based on the work of PR #47. First merge PR #47 (and then #50).

### Checklist

- [x] Have you added tests where necessary? Do all the test pass? 
- [x] Have you added descriptive comments to your code?
- [x] Have you updated the documentation related to this propo

### CODE REVIEW

File: src/movie/movie.resolver.ts [hunk 1]
```diff
@@ -1,16 +1,25 @@
...
 
   @Query()
-  async movie(@Args('imdbId') imdbId: string) {
+  async movie(@Args('imdbId') imdbId: string): Promise<Movie> {
     if (imdbId) {
-      return this.movieService.getMovieByImdbId(imdbId)
+      return await this.movieService.getMovieByImdbId(imdbId)
     } else {
-      return this.movieService.getRandomMovie()
+      return (await this.movieService.getRandomMovies())[0]
```
Comment (rschlaefli) [inline src/movie/movie.resolver.ts]: This will fail badly if we somehow don't get back what we expect.

File: src/movie/movie.resolver.ts [hunk 2]
```diff
@@ -1,16 +1,25 @@
...
 @Resolver('Movie')
 export class MovieResolver {
   constructor(private readonly movieService: MovieService) {}
 
   @Query()
-  async movie(@Args('imdbId') imdbId: string) {
+  async movie(@Args('imdbId') imdbId: string): Promise<Movie> {
     if (imdbId) {
-      return this.movieService.getMovieByImdbId(imdbId)
+      return await this.movieService.getMovieByImdbId(imdbId)
```
Comment (rschlaefli) [inline src/movie/movie.resolver.ts]: It is never necessary to "return await" as I think ESLint should also signify

File: src/movie/movie.resolver.ts [hunk 3]
```diff
@@ -1,16 +1,25 @@
...
+      return (await this.movieService.getRandomMovies())[0]
     }
   }
+
+  @ResolveField()
+  async randomMovies(
+    @Parent() movie: Movie,
+    @Args('num') num: number = 1,
+  ): Promise<Movie[]> {
+    return await this.movieService.getRandomMovies(num, movie.imdbId)
```
Comment (rschlaefli) [inline src/movie/movie.resolver.ts]: Remove all of the return await thingies

File: src/movie/movie.service.ts [hunk 4]
```diff
@@ -12,15 +12,8 @@ export class MovieService {
...
-      // get metadata
-      const metadataUrl = `${this.METADATA_SERVICE}?imdbId=${imdbId}`
-      const metadataResponse = await axios.get(metadataUrl)
-      const metadata = Utils.extractDataFromMetadataResponse(metadataResponse)
-
-      // get poster url
-      const posterUrl = `${this.POSTER_SERVICE}?id=${imdbId}`
-      const posterResponse = await axios.get(posterUrl)
-      const poster = Utils.extractDataFromPosterResponse(posterResponse)
+      const metadata = (await this.getMetadata(imdbId))[0]
```
Comment (rschlaefli) [inline src/movie/movie.service.ts]: While it is still not nice, this would be caught by the try/catch

File: src/movie/movie.service.ts [hunk 5]
```diff
@@ -33,26 +26,46 @@ export class MovieService {
...
-      const metadataResponse = await axios.get(metadataUrl)
-      const metadata = Utils.extractDataFromMetadataResponse(metadataResponse)
-
-      // get poster url
-      const posterUrl = `${this.POSTER_SERVICE}?id=${metadata.imdbId}`
-      const posterResponse = await axios.get(posterUrl)
-      const poster = Utils.extractDataFromPosterResponse(posterResponse)
+      const metadata = await this.getMetadata(null, numMovies, differentFrom)
+      const posters = await Promise.all(
+        metadata.map(async (m) => await this.getPoster(m.imdbId)),
```
Comment (rschlaefli) [inline src/movie/movie.service.ts]: Great ;) I also like to use Promise.all with map

Comment (rschlaefli) [inline src/movie/movie.service.ts]: Just as a hint, that should also work as `this.getPoster` already returns a Promise (you don't need to build your own):

```suggestion
        metadata.map(m => this.getPoster(m.imdbId)),
```

File: src/movie/movie.utils.ts [hunk 6]
```diff
@@ -1,14 +1,41 @@
...
+export const buildPosterServiceURL = (imdbId: string): string => {
+  return `${process.env.POSTER_SERVICE_URL}?id=${imdbId}`
+}
 
-  return {
-    imdbId: movie.imdb_id,
-    title: movie.title,
-    releaseYear: parseInt(movie.release_year),
-    posterPath: '',
+export const buildMetadataServiceURL = (
```
Comment (rschlaefli) [inline src/movie/movie.utils.ts]: Maybe we could use a URL query builder here to make it more extendable with different query params. Will need to look at such libraries.

**Comment by alexscheitlin** [comment]
Nice, thanks for the review! I'll have a look at it.

File: src/movie/movie.utils.ts (hunk 6 above)
Comment (rschlaefli) [reply src/movie/movie.utils.ts:7]: Check it out: https://nodejs.org/docs/latest-v12.x/api/url.html

Comment (rschlaefli) [reply src/movie/movie.utils.ts:7]: const newSearchParams = new URLSearchParams(myURL.searchParams);
// The above is equivalent to
// const newSearchParams = new URLSearchParams(myURL.search);

newSearchParams.append('a', 'c');
console.log(myURL.href);
// Prints https://example.org/?a=b
console.log(newSearchParams.toString());
// Prints a=b&a=c

File: src/movie/movie.resolver.ts (hunk 2 above)
Comment (alexscheitlin) [reply src/movie/movie.resolver.ts:12]: Fair enough (Y)

File: src/movie/movie.service.ts (hunk 5 above)
Comment (alexscheitlin) [reply src/movie/movie.service.ts:36]: Yes, I thought so too, but when I tried that it failed. However, now it works. It must have been something else. I'll adjust that accordingly.

File: src/movie/movie.service.ts (hunk 4 above)
Comment (alexscheitlin) [reply src/movie/movie.service.ts:15]: So what do you propose? Adding a separate try catch and throw a specific error message?

File: src/movie/movie.utils.ts (hunk 6 above)
Comment (alexscheitlin) [reply src/movie/movie.utils.ts:7]: Nice! A few weeks ago, I've read a medium article about that but forgot about it.

File: src/movie/movie.resolver.ts (hunk 1 above)
Comment (alexscheitlin) [reply src/movie/movie.resolver.ts:14]: That's correct. What would you suggest? Adding a try/catch block too? Or do you prefer then/catch?

Comment (rschlaefli) [reply src/movie/movie.resolver.ts:14]: Would be a nice addition. But I think we can also just leave it.

Comment (alexscheitlin) [reply src/movie/movie.resolver.ts]: Let's at least use then/catch :)

**Pull request merged**